Gas Analyzer can now scan pipes/devices along with the environment

theashtronaut opened the pull request:
<!-- The text between the arrows are comments - they will not be visible on your PR. -->
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->
The gas analyzer can now be used on atmospheric devices/pipes to display information on the GasMixtures contained within. It will also scan the tile the user is on regardless if an entity was scanned as well. When used, the analyzer will report all nodes in a NodeContainer. First though it will raise a GasAnalyzerScanEvent that atmospheric devices should subscribe to if they wish to override this behaviour or because they do not have a NodeContainerComponent. The gas analyzer will then display the contents in the UI. Support for up to Trinary is good, the UI will look bad if any devices with more than three mixes are scanned. Those don't exist yet though... The analyzer will shut off when you walk too far away from where you scanned as well. I also fixed the visual state so it changes when activated/deactivated.

I removed the old GasAnalyzable system entirely, GasAnalyzerSystem handles everything now.

Additionally, I tweaked ActivatableUISystem so that an item can keep it's UI up if it wants by adding the data field "closeOnHandDeselect". I desired this behavior for the analyzer as it supports working on pipes while still having it up and it felt wrong when it force closed the UI. I can see this having use with other items too like the PDA. The default is true so this preserves existing behavior. This behavior is demonstrated in the video.

As this is my first PR, please advise and correct anything I have done that is not kosher with how stuff should be done! I referenced a lot of different systems/components to mesh together the behavior I was after and modeled after those.

**Screenshots**
<!-- If applicable, add screenshots to showcase your PR. If your PR is a visual change, add
screenshots or it's liable to be closed by maintainers. -->


https://user-images.githubusercontent.com/112137107/188774312-a3318356-4cbc-4534-8f70-eac5d3be2a02.mp4


demo of functionality 

fixes space-wizards/space-station-14#8574 

**Changelog**
<!--
Here you can fill out a changelog that will automatically be added to the game when your PR is merged
There are 4 icons for changelog entries: add, remove, tweak, fix. I trust you can figure out the rest.

You can put your name after the :cl: symbol to change the name that shows in the changelog (otherwise it takes your GitHub username)
Like so: :cl: PJB

Generally, only put things in changelogs that players actually care about. Stuff like "Refactored X system, no changes should be visible" shouldn't be on a changelog.

For writing actual entries, don't consider the entry type suffix (e.g. add) to be "part" of the sentence:
bad: - add: a new tool for engineers
good: - add: added a new tool for engineers
-->

:cl: ashtronaut
- tweak: Nanotrasen has unlocked the gas analyzer's full potential and it can now be used to analyze atmospheric equipment

metalgearsloth commented on Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs:
```diff
@@ -1,46 +1,277 @@
     {
-        [Dependency] private readonly PopupSystem _popup = default!;
+        [Dependency] private readonly PopupSystem _popupSystem = default!;
+        [Dependency] private readonly IEntityManager _entityManager = default!;
```
There is already a ref to this on systems.

theashtronaut replied:
Oh I definitely missed that, removed the redundant reference. Thanks!

metalgearsloth commented on Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs:
```diff
@@ -1,46 +1,277 @@
     public sealed class GasAnalyzerSystem : EntitySystem
     {
-        [Dependency] private readonly PopupSystem _popup = default!;
+        [Dependency] private readonly PopupSystem _popupSystem = default!;
```
Don't put system in the field names as it's redundant.

theashtronaut replied:
Gotcha, renamed them all

mirrorcult commented on Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs:
```diff
@@ -1,46 +1,276 @@
+        [Dependency] private readonly SharedAppearanceSystem _appearance = default!;
+        [Dependency] private readonly UserInterfaceSystem _userInterface = default!;
+
+        private HashSet<EntityUid> _activeAnalyzers = new();
```
use a component to track these, something like `ActiveGasAnalyzerComponent`, and iterate over that with `EntityQuery` rather than keeping a hashset

theashtronaut replied:
I'll give this a shot, thanks!

mirrorcult commented on Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs:
```diff
@@ -1,46 +1,276 @@
         public override void Update(float frameTime)
         {
-            foreach (var analyzer in EntityManager.EntityQuery<GasAnalyzerComponent>(true))
+            if (!_activeAnalyzers.Any())
+                return;
```
this isnt necessary since the foreach will just skip over if it has none (and I believe this is incorrect since it should still be processing invalid analyzers?)

mirrorcult commented on Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs:
```diff
@@ -1,46 +1,276 @@
+                if (_invalidAnalyzers.List != null && _invalidAnalyzers.List.Contains(analyzer))
+                    continue;
+
+                if(!UpdateAnalyzer(analyzer))
+                    _invalidAnalyzers.Add(analyzer);
             }
+
+            foreach (var invalidAnalyzer in _invalidAnalyzers)
+            {
+                _activeAnalyzers.Remove(invalidAnalyzer);
+            }
+
+            _invalidAnalyzers.List?.Clear();
```
with the active comp/entityquery stuff mentioned above, the `invalidAnalyzers` bookkeeping can be removed entirely by just calling `RemCompDeferred<ActiveGasAnalyzerComponent>` in the entityquery

mirrorcult commented on Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs:
```diff
@@ -1,46 +1,276 @@
+            _activeAnalyzers.Add(uid);
+        }
+
+        private void OnComponentShutdown(EntityUid uid, GasAnalyzerComponent component, ComponentShutdown args)
+        {
+            // Make sure to remove this analyzer from the update list
+            _activeAnalyzers.Remove(uid);
+        }
```
no longer necessary if an `Active` comp is kept instead

mirrorcult commented on Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs:
```diff
@@ -1,46 +1,276 @@
                 _popup.PopupEntity(Loc.GetString("gas-analyzer-component-player-cannot-reach-message"), args.User, Filter.Entities(args.User));
                 return;
             }
+            ActivateAnalyzer(uid, component, args.User, args.Target);
+            OpenUserInterface(args.User, component);
```
is there any reason activating the analyzer is separate from opening the UI? useinhand activates the analyzer but doesn't open the UI, which seems like incorrect behavior

theashtronaut replied:
UseInHand opens the UI already due to having ActivatableUIComponent

mirrorcult replied:
ah okay

mirrorcult commented on Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs:
```diff
@@ -1,46 +1,276 @@
+            _activeAnalyzers.Remove(uid);
+        }
+
+        /// <summary>
+        /// Fetches new data when a user clicks on the refresh button
+        /// </summary>
+        private void OnRefreshMessage(EntityUid uid, GasAnalyzerComponent component, GasAnalyzerRefreshMessage message)
+        {
+            if (message.Session.AttachedEntity is not {Valid: true})
+                return;
+
+            UpdateAnalyzer(uid, component);
+        }
```
having an explicit refresh button is pointless if the gas analyzer updates automatically isn't it? it also doesn't look great in the UI, so removing that would be nice

theashtronaut replied:
Yeah this was a hold over from the prior implementation as it had a delay. But even when I add that delay, I agree, I'll remove it. Someone can always close/open the UI if for some reason it stops updating.

mirrorcult commented on Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs:
```diff
@@ -1,46 +1,276 @@
+                if (_invalidAnalyzers.List != null && _invalidAnalyzers.List.Contains(analyzer))
+                    continue;
+
+                if(!UpdateAnalyzer(analyzer))
```
this shouldn't update analyzers every tick--once per second or so should be good enough. there are lots of systems that do this kind of updating by just tracking accumulated frametime, e.g. `BloodstreamSystem`

mirrorcult commented on Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs:
```diff
@@ -1,46 +1,276 @@
+                return false;
+
+            // check if the user has walked away from what they scanned
+            var userPos = EntityManager.GetComponent<TransformComponent>(component.User).Coordinates;
```
```suggestion
            var userPos = Transform(component.User).Coordinates;
```

etc for other uses of this--you can also use `Comp<TransformComponent>`. these are just entity-specific shorthands

mirrorcult commented on Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs:
```diff
@@ -1,46 +1,276 @@
-                component.OpenInterface(actor.PlayerSession, args.ClickLocation);
+                // gas analyzed was used on an entity, try to request gas data
+                var ev = new GasAnalyzerScanEvent();
+                RaiseLocalEvent((EntityUid)component.Target, ev, false);
```
```suggestion
                RaiseLocalEvent(component.Target.Value, ev, false);
```

value-type nullability, more obvious what this is doing this way

mirrorcult commented on Content.Server/Atmos/EntitySystems/GasAnalyzerSystem.cs:
```diff
@@ -1,46 +1,276 @@
+                }
+                else
+                {
+                    // TODO: HACKY SOLUTION UNTIL I FIGURE OUT HOW TO HANDLE PIPES VIA EVENT
+                    // note though this will pick up any atmospherics devices that don't add a handler for the scan event
+                    // But they really should add that in themselves
```
this seems like a decent catch-all solution to me; as mentioned above maybe it should just default to this for all stuff? i suppose there could be a point to be made about not including specific gas mixtures if the system doesn't want them to be included, but you could solve that by reintroducing `GasAnalyzable` in some form

theashtronaut replied:
When I was first inquiring on this I was advised that an event system would be the way to go instead of using component references. I was able to do it for everything but pipes, hence this work around. But maybe I misunderstood what they meant?

mirrorcult replied:
oh--by component references i specifically meant the `[ComponentReference]` attribute + interfaces on the components, getting the nodecontainer is fine

theashtronaut replied:
Ohh gotcha, yeah I definitely misunderstood that. I can simplify it a bit in that case as any device that has a NodeContainer then would work fine. It would just leave gas tanks, canisters (it will report the Port not the mix), portables like the scrubber that would still have to rely on the event

mirrorcult commented on Content.Client/Atmos/UI/GasAnalyzerWindow.xaml.cs:
```diff
@@ -0,0 +1,197 @@
+    public sealed partial class GasAnalyzerWindow : DefaultWindow
+    {
+        public GasAnalyzerBoundUserInterface Owner { get; }
+        private TabContainer _tabContainer => CTabContainer;
+        private Button _refreshButton => RefreshButton;
+        private BoxContainer _topBox => CTopBox;
```
you shouldn't need to have these, you can just refer to `CTabContainer` etc. if you need them so that the BUI can access them, you can change access to public in XAML

mirrorcult reviewed (changes requested):
looks great. most major thing is the pipenode default behavior stuff i brought up

the UI might look better if the gas percentage color bar thing is anchored to the bottom of the tab container? not sure

theashtronaut commented on Content.Server/Atmos/Piping/Binary/EntitySystems/GasPassiveGateSystem.cs:
```diff
@@ -85,5 +86,23 @@ private void OnExamined(EntityUid uid, GasPassiveGateComponent gate, ExaminedEve
             args.PushMarkup(str);
         }
+
+        /// <summary>
+        /// Returns the gas mixture for the gas analyzer
+        /// </summary>
+        private void OnAnalyzed(EntityUid uid, GasPassiveGateComponent component, GasAnalyzerScanEvent args)
+        {
+            if (!EntityManager.TryGetComponent(uid, out NodeContainerComponent? nodeContainer))
+                return;
+
+            var gasMixDict = new Dictionary<string, GasMixture?>();
+
+            if(nodeContainer.TryGetNode(component.InletName, out PipeNode? inlet))
+                gasMixDict.Add(component.InletName, inlet.Air);
+            if(nodeContainer.TryGetNode(component.OutletName, out PipeNode? outlet))
+                gasMixDict.Add(component.OutletName, outlet.Air);
+
+            args.GasMixtures = gasMixDict;
+        }
```
Yeah this was all based on a prior discussion suggesting I made it event based instead of fetching component references. The 'default' is the fall-back for pipes since I couldn't figure out how to get them to react properly without resorting to component references.

Partmedia commented:
This looks nice!

Something that I have to do often is check the pressure on both nodes of a multi-node pipe. For example, just because the inlet of the pump is at 0 kPa doesn't necessarily mean something is wrong, because the output could be at a high pressure and the input side is just not keeping up. This is why I'd like it even more if I could see the pressures of all the nodes on one page, without having to tab. With automatic refresh, and multiple tabs, it might not even possible to see both refreshed at the same time.

So, how much work would it take to be able to see all the pressures on one page? Either 1) by having a summary tab with all the pressures and temperatures of each node at the front, or 2) just putting all of the nodes on one tab?

Partmedia commented:
Perhaps the "The gas analyzer shuts off" message every time the UI window is dismissed is not needed? No other user interface does this.

theashtronaut commented:
> So, how much work would it take to be able to see all the pressures on one page? Either 1) by having a summary tab with all the pressures and temperatures of each node at the front, or 2) just putting all of the nodes on one tab?

Yes I completely agree. My original vision had what 2 would be. I'll try to figure out a solution, I'm still learning all the UI stuff. The only tricky thing will be a potential loss of n-nary support (i.e. if devices are ever added with more than 3 nodes/GasMixtures) but I'm not sure how big of a deal that would be. I can also just throw excess nodes into tabs or something too.
No promises it'll be great, but if I can't figure it out I can at least add the summary info.

> Perhaps the "The gas analyzer shuts off" message every time the UI window is dismissed is not needed? No other user interface does this.

Yes, you are right that is an inconsistency and I will fix that so it only appears when the UI is force closed due to walking away or the item getting dropped.

theashtronaut commented:
Okay I hope I addressed all the prior review requests.

I also tweaked the UI. The original post has an updated video with the new UI. The auto resize gets a little wonk when going from trinary to unary but it's the best I could make it. Also it will properly report flipped devices, I just forgot to snag that on video.

I'm not super happy with the if, else if, else if combo in GasAnalyzerWindow.xaml.cs but I couldn't figure out a simpler solution since the panel used changes. I was always bad at those kinds of math problems.

Let me know what you think!

Partmedia commented:
I like how I can see all of the ports on one page, and they refresh at the same time. The addition of the picture of the entity being analyzed was also a great touch. A minor annoyance was that the UI window size kept automatically resizing, including when the text became longer.

Overall, I think gas analyzers are significantly better than before your PR. Thank you!

theashtronaut commented:
> A minor annoyance was that the UI window size kept automatically resizing, including when the text became longer.

I made that change because I personally though it was annoying to have to keep resizing the window myself, especially when switching from trinary, to binary, to unary, or any combination between.

For reference here is the new UI with no auto resize at all

https://user-images.githubusercontent.com/112137107/188295401-34a61640-f711-493d-87a6-755fb5d16e9a.mp4



I also tested only resizing if the desired size is bigger, but not auto resizing to be smaller. This would mean that when the text gets longer it will still expand, but it wouldn't fluctuate if it was going smaller->bigger->smaller

here is what that looks like:

https://user-images.githubusercontent.com/112137107/188295431-5a78cda0-9fbd-4b94-9d67-c9c247fc1dd0.mp4

If there are really strong feelings on if I should completely remove auto resize, or switch it to only make things larger, definitely voice them!

Partmedia commented:
This is what I see:

https://user-images.githubusercontent.com/3229565/188296014-bcbf5e6a-eb3b-4eac-bf23-ef6bbc641301.mp4

Note that the borders of the UI dialog seem to be constantly repacking themselves.

theashtronaut commented:
Huh that is very strange. Is it doing that for other devices for you?

Did you add an event handler for GasAnalyzerScanEvent for the gas recycler? I don't that that in my branch.

I was able to.. sorta barely replicate it when hooking up a ThermoMachine to some pipes, but it was only going mildly wild some of the time. Not nearly as bad as your example where every update it was changing.

I'll try to dig into this further and see if I can figure out why it might be happening for sure.

Partmedia commented:
I don't override the `GasAnalyzerScanEvent`. Let me see if I can produce a minimum example to reproduce (or at least send you a map file with which branch to check out) for you to reproduce locally.

theashtronaut commented:
> I don't override the `GasAnalyzerScanEvent`. Let me see if I can produce a minimum example to reproduce (or at least send you a map file with which branch to check out) for you to reproduce locally.

Okay, it's only needed if the default doesn't work for the device. I only asked because I was only able to see that behaviour on the ThermoMachine which does handle the event. I'll keep trying in the mean time.

Would also love to hear if any others testing it see it as well

theashtronaut commented:
Okay, good news I was able to reproduce what you were seeing. The error was in my testing methodology :stuck_out_tongue_closed_eyes: 

I tweaked the UI to hopefully alleviate this issue. I made all the value labels right set, and set the middle separator control along with the value labels to expand. I also increased the minimum size for the window/data containers. In my testing with 1000s of mols and temperature changing it didn't auto-resize and there is enough room for hundreds of millions of mols before it starts to do the jank. I tested with CO2 since that was the longest gas name currently. This isn't technically a "permanent" fix but more of a "you shouldn't see this often in regular usage" so I'm sure it's fine. Downside is the UI is a little big now though. 

Again if desired, I can just remove the auto-resize logic entirely and return the window to being smaller. Just force everyone to manually re-adjust the window and all. Lmk what you think!

mirrorcult commented:
couple things to note in testing

- the autoresize seems to be delayed by a frame on window open, you can kinda see in this video that it starts out at max size and then next frame shrinks which looks really strange


https://user-images.githubusercontent.com/19853115/188535528-fc26f5d7-2581-4a43-a97b-66e2ff54ad64.mp4

- non-filter trinary devices are incorrectly labeled. this is a mixer, but the 2nd inlet is labeled 'filter'

![image](https://user-images.githubusercontent.com/19853115/188535698-246c38fa-55b5-4b52-90f4-ad8ee0c299f2.png)

- flipped trinary devices are also labeled incorrectly

![image](https://user-images.githubusercontent.com/19853115/188535875-d32bea4e-e9e7-48e8-80d4-29d627a96f0a.png)

theashtronaut commented:
> couple things to note in testing
> 
>     * the autoresize seems to be delayed by a frame on window open, you can kinda see in this video that it starts out at max size and then next frame shrinks which looks really strange

This will always be sort of an issue, but I changed it so instead it will start smaller instead of bigger so it has more of an 'opening' frame instead of closing. Hopefully that's better. The reason why is because the auto resize relies on getting the data, which just takes a frame or so to get back. Probably worse if there is high latency.

> 
>     * non-filter trinary devices are incorrectly labeled. this is a mixer, but the 2nd inlet is labeled 'filter'
> 

It uses the name of the NodeContainer entry. Here's two mixers in Saltern:

![image](https://user-images.githubusercontent.com/112137107/188545557-e60f8e24-bb0f-452a-96c0-796de37c2fd5.png)

![image](https://user-images.githubusercontent.com/112137107/188545881-dabfbbca-f660-4337-b435-d08afe48098b.png)

I'm not actually sure why, perhaps some mapping thing? The defaults for GasMixerComponent are inletOne and inletTwo. Which actually would look pretty bad as a title, I might want go back to having mixers and filters override the event and set keys for them.



> 
>     * flipped trinary devices are also labeled incorrectly
> 

Yeah, I'll add some logic to check for that and fix that!

Why did it fail integration tests on github? It passed when I ran them before pushing.

mirrorcult commented:
test fail is erroneous

theashtronaut commented:
Okay! I made the changes I talked about.

The data boxes in the UI start as not visible, so for the first frame or so instead of being max size it's smaller. This gives that frame or two a feeling of an "opening" animation instead of a "closing" animation. 

I manually have Filters and Mixers setting the names for the nodes.

I added proper support for flippable trinary devices now. I updated the video in the main post showing these.

I also fixed the item animation not working after reviewing some other things. That's been broken this entire time but wasn't really critical.

mirrorcult commented on Resources/Prototypes/Entities/Objects/Specific/atmos.yml:
```diff
@@ -20,10 +23,10 @@
-        analyzer:
-          Off: { state: icon }
-          Working: { state: working }
+        enum.FilterVisuals.Enabled:
```
using filtervisuals for this seems strange, just make another enum

theashtronaut replied:
For some reason it didn't work the first time I tried, but I was definitely doing something wrong. it's got it's own enum now

mirrorcult commented on Content.Server/Atmos/Piping/Trinary/EntitySystems/GasFilterSystem.cs:
```diff
@@ -159,5 +160,31 @@ private void OnSelectGasMessage(EntityUid uid, GasFilterComponent filter, GasFil
+                gasMixDict.Add(Loc.GetString("gas-analyzer-window-text-filter"), filterNode.Air);
+            if(nodeContainer.TryGetNode(component.OutletName, out PipeNode? outlet))
+                gasMixDict.Add(Loc.GetString("gas-analyzer-window-text-outlet"), outlet.Air);
+
+
```
```suggestion
```

mirrorcult commented on Content.Server/Atmos/Piping/Trinary/EntitySystems/GasMixerSystem.cs:
```diff
@@ -203,5 +204,38 @@ private void OnOutputPressureChangeMessage(EntityUid uid, GasMixerComponent mixe
+                flipped = false;
+            */
+            args.GasMixtures = gasMixDict;
+            args.DeviceFlipped = inletOne != null && inletTwo != null && inletOne.CurrentPipeDirection.ToDirection() == inletTwo.CurrentPipeDirection.ToDirection().GetClockwise90Degrees();
```
i definitely feel like theres still a better way to do this but eh its good

theashtronaut replied:
probably, but it's the best I could come up with after doodling it all out. It's an arbitrary convention to fit the UI for sure

open to suggestions if someone thinks of a better way

theashtronaut commented:
The only lasting thing I could never figure out was why tooltips for the fancy gas bar don't work, but I don't think it's worth blocking over since the information is duplicated in the display anyway.
